File: 221 Maximal Square/221 Maximal Square.go
func maximalSquare(matrix [][]byte) int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return 0
	}
	// row column size
	r, c := len(matrix), len(matrix[0])
	res := 0
	dp := make([][]int, r)
	for i := range dp {
		dp[i] = make([]int, c)
	}
	for i := 0; i < r; i++ {
		for j := 0; j < c; j++ {
			if i == 0 || j == 0 {
				dp[i][j] = int(matrix[i][j] - '0')
			} else if matrix[i][j] == '1' {
				dp[i][j] = Min(dp[i-1][j-1], Min(dp[i][j-1], dp[i-1][j])) + 1
			}
			res = Max(res, dp[i][j])
		}
	}

	return res * res
}

func Min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 221 Maximal Square/221 Maximal Square.py
"""
Given an m x n binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

Example 1:

Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
Output: 4

Example 2:

Input: matrix = [["0","1"],["1","0"]]
Output: 1

Example 3:

Input: matrix = [["0"]]
Output: 0
"""
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        area = 0
        if matrix:
            prev = [0]*len(matrix[0])
            for row in matrix:
                s = [1 if ch == '1' else 0 for ch in row]
                for j in range(1,len(s)):
                    #if s[j] = 0 then still = 0 if not, we update
                    if s[j]:
                        s[j] = min(prev[j-1],prev[j],s[j-1])+1
                area = max(area,max(s)**2)
                prev = s
        return area
